jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/internal/ContentCopier.java

```diff
@@ -46,43 +47,38 @@ public InvocationType getInvocationType()
     @Override
     protected Action process() throws Throwable
     {
+        if (current != null)
+            current.release();
+
         if (terminated)
             return Action.SUCCEEDED;
 
         current = source.read();
 
         if (current == null)
         {
-            source.demand(this::iterate);
-            return Action.IDLE;
+            source.demand(this::succeeded);
+            return Action.SCHEDULED;
         }
 
         if (chunkProcessor != null && chunkProcessor.process(current, this))
             return Action.SCHEDULED;
 
+        terminated = current.isLast();
+
         if (Content.Chunk.isFailure(current))
-            throw current.getFailure();
+        {
+            failed(current.getFailure());
+            return Action.SCHEDULED;
```

Comment:
Currently throwing here is bad, as the ICB will call failed, but then throw an ISE because the action is null.
So calling failed(x) and then returning SCHEDULED is simulating an async call that failed within the scope of the call.    The updated ICB in the other branch is more robust for null actions. 

The current should be released, but I don't think we have any failure chunks that actually do anything on release.